### PR TITLE
Add missing xorg dependencies

### PR DESCRIPTION
**Pull request by johanneskoester**
This fixes issue #18.

### CODE REVIEW

**Comment by ccordoba12** [comment]
@conda-forge-admin, please rerender

**Comment by johanneskoester** [comment]
Looks like they don't provide 2.38 sources anymore... I guess we should update to 2.40 then.

**Comment by jakirkham** [comment]
For Windows:
* https://graphviz.gitlab.io/_pages/Download/windows/graphviz-2.38.zip

For Linux:
* https://gitlab.com/graphviz/graphviz/commit/f54ac2c9313ae80ccf76ef4ac6aa9be820a23126 (from commit?)
* http://pkgs.fedoraproject.org/repo/pkgs/graphviz/graphviz-2.38.0.tar.gz/5b6a829b2ac94efcd5fa3c223ed6d3ae/graphviz-2.38.0.tar.gz (from mirror?)
* https://gitlab.com/graphviz/graphviz/issues/1350 (ask for help?)

File: recipe/meta.yaml [hunk 1]
```diff
@@ -25,26 +22,28 @@ build:
...
+    - libtiff                  # [unix]
     - libtool                  # [linux]
-    - freetype 2.8.1             # [unix]
-    - jpeg 9*                  # [unix]
-    - pango 1.40.*             # [linux]
-    - zlib 1.2.11               # [unix]
+    - freetype                 # [unix]
+    - jpeg                     # [unix]
+    - pango                    # [linux]
+    - zlib                     # [unix]
```
Comment (ccordoba12) [inline recipe/meta.yaml]: This is really bad! Please restore all the pinnings here.

Why did you remove them?

Comment (johanneskoester) [reply recipe/meta.yaml]: I removed them because we have the same pinnings in the pinning feedstock.

Comment (ccordoba12) [reply recipe/meta.yaml]: > in the sense that it would be incompatible with other versions

That's the main reason those pinnings are in place, e.g. to avoid incompatibilities with different versions of freetype, jpeg, etc.

@jakirkham?

File: recipe/meta.yaml [hunk 2]
```diff
@@ -20,38 +17,47 @@ source:
...
   build:
-    - cairo 1.14.*             # [linux]
-    - expat 2.2.*                   # [linux]
-    - libpng >=1.6.32,<1.6.35  # [unix]
-    - libtiff >=4.0.8,<4.0.10   # [unix]
-    - libtool                  # [linux]
-    - freetype 2.8.1             # [unix]
-    - jpeg 9*                  # [unix]
-    - pango 1.40.*             # [linux]
+    - {{ compiler('cxx') }}    # [not osx]
```
Comment (ccordoba12) [inline recipe/meta.yaml]: Nothing is compiled on Windows. The binaries provided by Graphviz are just repackaged as a conda package, so this makes no sense.

Comment (johanneskoester) [reply recipe/meta.yaml]: I know, but removing it caused it to fail on win (see [here](https://ci.appveyor.com/project/conda-forge/graphviz-feedstock/build/1.0.42)). I don't know why either. But it also does not really hurt, because we might want to compile in the future.

File: recipe/meta.yaml (hunk 1 above)
Comment (johanneskoester) [reply recipe/meta.yaml]: What I mean is: if a package needs a certain version because of the API, it should be pinned in the dependencies. If a package needs to pin just to maintain ABI compatibility (the runtime version has to be compatible to the build time version) this can be handled by CB3 via the pinning feedstock.

Comment (johanneskoester) [reply recipe/meta.yaml]: By the pinning feedstock, it would even be possible to build grahpviz for different versions of e.g. freetype (which we don't do though). Also, updating the pinning in the feedstock means that all packages can be consistently updated to a new pinning, such that it is easier to combine them in the same environment.

Comment (johanneskoester) [reply recipe/meta.yaml]: Also see [the docs](https://conda.io/docs/user-guide/tasks/build-packages/variants.html)

**Comment by johanneskoester** [comment]
Mhm, I have no idea what's wrong with osx though: https://travis-ci.org/conda-forge/graphviz-feedstock/builds/397188535#L2296

**Comment by jakirkham** [comment]
@conda-forge-admin, please re-render.

File: recipe/meta.yaml (hunk 1 above)
Comment (jakirkham) [reply recipe/meta.yaml]: Johannes is correct.

We migrated what was in the old pinning script to that pinning file. If you look at the dependencies here and that pinning file, they should be pretty close (if not exact) matches to what we have here. A few shifted slightly over time to address issues that came up in the beginning.

Re-rendering pulls a subset of those pinnings into the feedstock (just did that now). So one can see how these will be pinned and verify they are correct.

Happy to answer any questions. :)

**Comment by jakirkham** [comment]
The macOS issue seems to be some sort of issue with `libuuid` that is getting pulled in. Likely through the `xorg-*` dependencies. Maybe we can restrict those to Linux for now?

xref: https://github.com/Homebrew/legacy-homebrew/issues/32587 (similar issue in Homebrew)

File: recipe/meta.yaml (hunk 1 above)
Comment (ccordoba12) [reply recipe/meta.yaml]: Ok, I see, thanks for the clarification @johanneskoester and @jakirkham.

But this is a feature of conda-smithy 3, paired with conda-build 3, right? I mean, it sounded strange to me when @johanneskoester said that pinnings are auto-magically handled by conda-build only.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -20,38 +17,48 @@ source:
...
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
     - pkg-config               # [osx]
-    - zlib 1.2.11               # [unix]
+  host:
+    - cairo                    # [linux]
+    - expat                    # [linux]
+    - libpng                   # [unix]
+    - libtiff                  # [unix]
+    - libtool                  # [linux]
```
Comment (jakirkham) [inline recipe/meta.yaml]: Maybe we should start using this on macOS too? Looks like the `libtool` being used on macOS is running into [some issues]( https://travis-ci.org/conda-forge/graphviz-feedstock/builds/397342300 ).

File: recipe/meta.yaml (hunk 1 above)
Comment (jakirkham) [reply recipe/meta.yaml]: Sure there is some work being done in the background by our tooling.

In the future, we may move to [`run_exports`]( https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#export-runtime-requirements ), which would avoid needing this work done by `conda-smithy`. Example usage can be seen with [`zlib`]( https://github.com/conda-forge/zlib-feedstock/blob/fa4368510c67a2dd14c19e0ed2791bd53ed958bb/recipe/meta.yaml#L14-L17 ).

File: recipe/build.sh [hunk 4]
```diff
@@ -1,6 +1,8 @@
 #!/bin/bash
 
 if [ `uname` == Darwin ]; then
+    export OBJC="${CC}"
```
Comment (jakirkham) [inline recipe/build.sh]: Turns out `libtool` didn't know what `OBJC` was and that is [what caused issues before]( https://travis-ci.org/conda-forge/graphviz-feedstock/builds/397490964#L2235 ). The solution is to explain what compiler is needed for `OBJC`, which turns out to be the same as `CC` (a.k.a. `clang`) on macOS. With that change alone, [the build passes]( https://travis-ci.org/conda-forge/graphviz-feedstock/builds/400719599 ). Not sure why we didn't need this change before. 🤔

**Comment by jakirkham** [comment]
@conda-forge-admin, please re-render.

File: recipe/build.sh (hunk 4 above)
Comment (jakirkham) [reply recipe/build.sh]: Raised issue ( https://github.com/conda-forge/toolchain-feedstock/issues/40 ) about getting this into the `compiler` syntax in the future.

File: recipe/meta.yaml [hunk 5]
```diff
@@ -1,17 +1,14 @@
...
 
 source:
-  fn:  graphviz-{{ version }}.tar.gz  # [unix]
-  url: http://www.graphviz.org/pub/graphviz/stable/SOURCES/graphviz-{{ version }}.tar.gz  # [unix]
+  url: http://pkgs.fedoraproject.org/repo/pkgs/graphviz/graphviz-2.38.0.tar.gz/5b6a829b2ac94efcd5fa3c223ed6d3ae/graphviz-2.38.0.tar.gz  # [unix]
   md5: 5b6a829b2ac94efcd5fa3c223ed6d3ae  # [unix]
 
-  fn:  graphviz-{{ win_version }}.zip  # [win]
-  url: http://www.graphviz.org/pub/graphviz/stable/windows/graphviz-{{ win_version }}.zip  # [win]
+  url: https://graphviz.gitlab.io/_pages/Download/windows/graphviz-2.38.zip  # [win]
```
Comment (jakirkham) [inline recipe/meta.yaml]: Would be nice to template the URLs to make them a little nicer for use in the future.

File: recipe/meta.yaml [hunk 6]
```diff
@@ -1,17 +1,14 @@
...
 
 package:
   name: graphviz
   version: {{ version }}
 
 source:
-  fn:  graphviz-{{ version }}.tar.gz  # [unix]
-  url: http://www.graphviz.org/pub/graphviz/stable/SOURCES/graphviz-{{ version }}.tar.gz  # [unix]
+  url: http://pkgs.fedoraproject.org/repo/pkgs/graphviz/graphviz-2.38.0.tar.gz/5b6a829b2ac94efcd5fa3c223ed6d3ae/graphviz-2.38.0.tar.gz  # [unix]
   md5: 5b6a829b2ac94efcd5fa3c223ed6d3ae  # [unix]
```
Comment (jakirkham) [inline recipe/meta.yaml]: Note the URL above is just using `md5`. So this can be templated out easily.

File: recipe/meta.yaml [hunk 7]
```diff
@@ -20,38 +17,48 @@ source:
...
+    - xorg-libxrender          # [linux]
+    - xorg-libxpm              # [linux]
 
 test:
   commands:
     - dot -V
     - neato -?  # [unix]
     - conda inspect linkages $PKG_NAME  # [not win]
     - conda inspect objects $PKG_NAME  # [osx]
+    - dot -Tsvg sample.dot
```
Comment (jakirkham) [inline recipe/meta.yaml]: Maybe we can move `svg` into `run_test.sh` and `run_test.bat`?

Comment (johanneskoester) [reply recipe/meta.yaml]: You mean that last line with the `dot` command? May I ask why?

File: recipe/meta.yaml (hunk 6 above)
Comment (johanneskoester) [reply recipe/meta.yaml]: That is true, but for the next release we will simply have a gitlab URL which only contains the version and not the md5. Hence, I would argue for keeping it like this for now.

Comment (jakirkham) [reply recipe/meta.yaml]: My concern is the autoupdate bot will come through and create a PR that changes the version up top in the Jinja, but the URL will stay the same. If that passes, someone could end up merging this not realizing the URLs are incorrect. This has happened before. ( https://github.com/conda-forge/sqlite-feedstock/issues/25 ) Hence the request to fix up the URLs. Am ok if it doesn't generalize and does fail on the next release as that will cause someone to look into it instead of potentially overlooking it and merging because the CI passed for something that is incorrect.

File: recipe/meta.yaml (hunk 7 above)
Comment (jakirkham) [reply recipe/meta.yaml]: This appears to already be tested. Should we have a better test?

ref: https://github.com/conda-forge/graphviz-feedstock/blob/046c1d35d8c8f6042d253d74220a323c1c5b388b/recipe/run_test.sh#L5
ref: https://github.com/conda-forge/graphviz-feedstock/blob/046c1d35d8c8f6042d253d74220a323c1c5b388b/recipe/run_test.bat#L1

File: recipe/meta.yaml (hunk 6 above)
Comment (johanneskoester) [reply recipe/meta.yaml]: Ah, that makes sense, I see. Will do that. Suggestion: let the autoupdate bot check whether URL changes after changing the version.

**Comment by johanneskoester** [comment]
@jakirkham I think this is now ready for merge if you agree.

**Comment by johanneskoester** [comment]
Could anybody of the maintainers merge this? Thanks a lot!

File: recipe/meta.yaml [hunk 8]
```diff
@@ -1,17 +1,14 @@
 {% set version = "2.38.0" %}
```
Comment (ccordoba12) [inline recipe/meta.yaml]: Please add

`{% set name = "graphviz" %}`

on top of this file.

File: recipe/meta.yaml [hunk 9]
```diff
@@ -1,17 +1,14 @@
...
-{% set win_version = version[:-2] %}
 
 package:
   name: graphviz
   version: {{ version }}
 
 source:
-  fn:  graphviz-{{ version }}.tar.gz  # [unix]
-  url: http://www.graphviz.org/pub/graphviz/stable/SOURCES/graphviz-{{ version }}.tar.gz  # [unix]
+  url: http://pkgs.fedoraproject.org/repo/pkgs/graphviz/graphviz-{{ version }}.tar.gz/5b6a829b2ac94efcd5fa3c223ed6d3ae/graphviz-{{ version }}.tar.gz  # [unix]
```
Comment (ccordoba12) [inline recipe/meta.yaml]: Please use `{{ name }}` here to replace the instances of `graphviz`.

File: recipe/meta.yaml [hunk 10]
```diff
@@ -1,17 +1,14 @@
...
   md5: 5b6a829b2ac94efcd5fa3c223ed6d3ae  # [unix]
 
-  fn:  graphviz-{{ win_version }}.zip  # [win]
-  url: http://www.graphviz.org/pub/graphviz/stable/windows/graphviz-{{ win_version }}.zip  # [win]
+  url: https://graphviz.gitlab.io/_pages/Download/windows/graphviz-{{ version[:-2] }}.zip  # [win]
```
Comment (ccordoba12) [inline recipe/meta.yaml]: Also use `{{ name }}` here.

File: recipe/meta.yaml (hunk 8 above)
Comment (ccordoba12) [reply recipe/meta.yaml]: Ok, merging then. Thanks @johanneskoester for your help!

**Comment by jakirkham** [comment]
Thanks @johanneskoester and @ccordoba12 :)
